src/Microsoft.VisualStudio.ProjectSystem.Managed/ProjectSystem/LanguageServices/ApplyChangesToWorkspaceContext.cs

```diff
@@ -69,11 +73,29 @@ public async Task ApplyProjectBuildAsync(IProjectVersionedValue<IProjectSubscrip
             {
                 IComparable version = GetConfiguredProjectVersion(update);
 
+                ProcessOptions(buildSnapshot.Value);
                 await ProcessCommandLineAsync(version, projectChange.Difference, state, cancellationToken);
                 ProcessProjectBuildFailure(projectChange.After);
             }
         }
 
+        private void ProcessOptions(IProjectBuildSnapshot buildSnapshot)
+        {
+            Assumes.NotNull(_context);
+
+            List<string> commandLineArguments = new List<string>();
+
+            buildSnapshot.TargetOutputs.TryGetValue("CompileDesignTime", out var keyValuePairsCommandLineOptions);
+
+            foreach (var kp in keyValuePairsCommandLineOptions)
+            {
+                commandLineArguments.Add(kp.Key);
+            }
+
+            // We just need to pass all options to Roslyn
+            _context.SetOptions(commandLineArguments.ToImmutableArray());
```

Comment:
To reduce allocations (written in a browser, so untested):

```suggestion
            buildSnapshot.TargetOutputs.TryGetValue(
                "CompileDesignTime",
                out IImmutableList<KeyValuePair<string, IImmutableDictionary<string, string>>> targetOutputs);

            var options = ImmutableArray.CreateBuilder<string>(targetOutputs.Count);

            foreach ((string option, _) targetOutputs)
            {
                options .Add(option);
            }

            // We just need to pass all options to Roslyn
            _context.SetOptions(options .MoveToImmutable());
```

By using a builder with a predefined capacity, and adding exactly that number of items, we can use `MoveToImmutable` which moves the backing array from the builder to the `ImmutableArray<string>`. This means there is only a single array allocated during this operation.

You need to handle the case where `TryGetValue` returns false, as then `targetOutputs` will be `null`, resulting in an exception.

Reply:
If you know the size beforehand, what is the advantage of using an `ImmutableArray<T>.Builder` over just using a normal array, populating it, and then calling `ImmutableArray<T>.Create()`? It seems more straightforward that way.

Reply:
@MiYanni I never replied.

`ImmutableArray<T>.Create(T[])` will copy the contents of the passed array to a new array, resulting in extra work and allocation. It does this copy so that you cannot modify the contents of the array. `ImmutableArray<>` is a wrapper around an array, and it is designed to make it very hard to modify the contents of the array, as the type's name would suggest.

The builder can encapsulate the array, so no one outside can modify it without its knowledge (via its API). The `MoveToImmutable` method "moves" the builder's array directly into the `ImmutableArray<>`.